[HIBIKE] Some unit tests

baby-bell opened the pull request:
Previously, Hibike had no unit tests, although some testing infrastructure was in place (`hibike_tester` and `virtual devices`). This commit adds a few, nonexhaustive unit tests for functions in `hibike_message` and `hibike_process`. As of now, most tests are based on randomized input data, rather than specific inputs.

baby-bell committed:
Add unit tests for cobs

baby-bell committed:
Add unit testing module

Add random device data and sub response

Add test for parse idempotence

Add more devices to virtual_device

Add verbose mode to virtual_device

Add simple reading test for blocking_read

Test parsing bad checksum packets

Tests for params_encode/decode

Test device identify for empty ports mixed with devices

Previously, Hibike had no unit tests, although some testing
infrastructure was in place (hibike_tester and virtual devices).
This commit adds a few, nonexhaustive unit tests for
functions in hibike_message and hibike_process. As of
now, most tests are based on randomized input data, rather
than specific inputs.

baby-bell commented:
These tests can be added to Travis, if preferred.

tingtingchiang commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,385 @@
+
+import hibike_message
+import spawn_virtual_devices
+DEVICE_TYPES = [dev_id for dev_id in hibike_message.DEVICES]
```
Maybe just a stylistic choice, but list(hibike_message.DEVICES) reads better for me, or hibike_message.DEVICES.keys(). Both create a new list of keys.

baby-bell replied:
Can do!

tingtingchiang commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,385 @@
+    for _ in range(times):
+        # Do not expand list arguments out
+        generated_args = arg_func()
+        if isinstance(generated_args, tuple):
```
Does generated_args change from being a tuple to another type of instance between loops? If not, we should just generated the arguments and check once before deciding how to call the func. Otherwise, this for loop may become very expensive if arg_func is expensive.

baby-bell replied:
Yeah, I think you're right about this, especially since the data we are generating should always be of the same type.

tingtingchiang reviewed (changes requested):
Is it possible to also add a README file with descriptions on how to run tests? Or is there an example of how to run them somewhere?

baby-bell committed:
Fix Tiffany issues

baby-bell committed:
Fix virtual device verbose mode

baby-bell commented:
Just a warning: the waiting period for virtual devices is not enough for running tests on the Beaglebone, so it will usually fail `hibike_process_tests`. I can modify this if you like.

yizow commented on hibike/unit_tests/hibike_process_tests.py:
```diff
@@ -0,0 +1,65 @@
+    """
+    Tests for `identify_smart_sensors`.
+    """
+    VIRTUAL_DEVICE_TYPES = ["LimitSwitch", "YogiBear", "YogiBear",
```
Will this ever be used / synced with something else? If so, it should be in a central file that all things are importing.

baby-bell replied:
This is just an example set of sensors that might be put on a robot. If there's anything that might be factored out, it would be `run_with_random_data`

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,386 @@
+
+import hibike_message
+import spawn_virtual_devices
+DEVICE_TYPES = list(hibike_message.DEVICES)
```
Why convert this to a list? `hibike_message.devices` is originally a dictionary that should have all the info you need. The way youre using `DEVICE_TYPES` later also suggests you want access to that data

baby-bell replied:
`random.choice`, which I am using a few times, only works with sequences. This doesn't include dictionaries.

kevinji replied:
You could probably just use `hibike_message.DEVICES.keys()` where necessary, to avoid instantiating a copy of the keys.

baby-bell replied:
I'm afraid that still doesn't work. `dict.keys` gives an iterator, not a sequence.

yizow commented on hibike/hibike_message.py:
```diff
@@ -405,17 +415,18 @@ def parse_bytes(msg_bytes):
+    while not stop_event.is_set():
         # Wait for a 0 byte to appear
-        while packets_buffer.find(zero_byte) == -1:
+        while packets_buffer.find(zero_byte) == -1 and not stop_event.is_set():
```
why are we duplicating this is_set here?
If this is an edge case that causes problems, then you should be able to restructure the code to address that logic.

baby-bell replied:
The goal is to be able to stop reads when the event is set. It's possible to get stuck in this loop if we don't check the event.

kevinji replied:
You can change this into an `if` and stick a `continue` at the end of this `while` loop to avoid repeating the `not stop_event.is_set()` condition.

baby-bell replied:
I'm sorry, I don't quite understand what you mean by that. Which part am I changing into an `if`?

kevinji replied:
I was imagining something like the following:
```python
while not stop_event.is_set():
    # Wait for a 0 byte to appear
    if packets_buffer.find(zero_byte) == -1:
        # ...
        continue

    # ...
```
Although I'm not sure what the correct behavior should be if a stop event has occurred. Should the last "read" packet still be `yield`-ed, like the current code does?

baby-bell replied:
It's OK to stop immediately when the event is set, rather than yield additional packets. Once calling code sets the event, it's assumed that it doesn't care about any other data. I'll try using this fix, thanks.

yizow commented on hibike/hibike_message.py:
```diff
@@ -104,6 +106,14 @@ def __str__(self):
+    def __eq__(self, other):
+        if not isinstance(other, HibikeMessage):
+            return False
+        return self._message_id == other._message_id and\
```
For multi line, preferred style is implicit continuation via wrapping in parentheses, and putting the boolean operator at the start of the line. 

see [here](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator)

yizow commented on hibike/hibike_message.py:
```diff
@@ -104,6 +106,14 @@ def __str__(self):
+        if not isinstance(other, HibikeMessage):
+            return False
+        return self._message_id == other._message_id and\
+               self._length == other._length and\
```
does a `HibikeMessage` have a is_valid check or similar? If so, no need to compare the length with the payload

yizow replied:
And if not, maybe its time to write one : P

yizow commented on hibike/hibike_message.py:
```diff
@@ -437,10 +448,13 @@ def blocking_read_generator(serial_conn, stop_event=None):
             # Otherwise, there might be more incoming bytes for the current packet,
             # so we do a blocking read and try again
-            else:
+            elif not stop_event.is_set():
```
i feel like putting this check *everywhere* is masking some underlying code structure issue. if you're worried about synchronization, then its a great time to
1) think hard about exactly how the threads/ processes can interleave and interfere
2) design the best, easiest, cleanest way to solve it
3) learn about python multi-threading/processing synchronization primitives

baby-bell replied:
Actually, this particular check is not necessary anyways, since reads are nonblocking. I will remove that.

baby-bell committed:
Test lint failure

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,386 @@
+
+def test_with_random_data(func, arg_func, kwarg_func=lambda: {}, times=5, *args, **kwargs):
+    """
+    Test FUNC with random arguments generated by ARG_FUNC and KWARG
```
i dont like changing the case of variable names, especially in a case-sensitve language. 

also `/s/KWARG\nfunc/KWARG_FUC`

yizow replied:
I'd also replace "Test" with "Execute" or "Run"

baby-bell replied:
Doesn't look like there's ambiguity wrt variable names here.

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,386 @@
+    func, repeating TIMES.
+    """
+    # Do not expand returned lists out
+    multiple_args = isinstance(arg_func(), tuple)
```
Do you require arg_func() to be idempotent?

baby-bell replied:
No, it is supposed to generate random data.

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,386 @@
+    """
+    # Do not expand returned lists out
+    multiple_args = isinstance(arg_func(), tuple)
+    if multiple_args:
```
get rid of the if-branch:
`multiple_args = tuple(arg_func())`

Or, just specify that arg_func must return an iterable in your docstring

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,386 @@
+    Generate a random UID.
+    """
+    uid_type = random.choice(DEVICE_TYPES)
+    return uid_type << 72 | random.getrandbits(8) << 64 | random.getrandbits(64)
```
why not:
`uid_type << 72 |  random.getrandbits(72)`

baby-bell replied:
Good point.

yizow commented on hibike/virtual_device.py:
```diff
@@ -89,19 +105,22 @@ def main():
             # Update the delay, subscription time,
             # and params, then send a subscription response
-            print("Subscription request recieved")
+            if args.verbose:
```
Hmm i feel like we could restructure this to get rid of arg.verbose check multiple times.
Eh probably a future pull request

kevinji replied:
We can just have a function called `log` that contains the `args.verbose` logic, and then use `log("MESSAGE")` instead.

yizow commented on hibike/virtual_device.py:
```diff
@@ -26,11 +26,15 @@ def main():
     port = args.port
-    print(device, port)
+    if args.verbose:
+        print(device, port)
```
Add a more descriptive string

yizow reviewed (changes requested):
Will finish review of more in depth additional classes, etc. later

Also I'm gonna block till i get a good commit message : P

baby-bell committed:
Disable mysterious failing lint

baby-bell committed:
Disable lint for real

baby-bell committed:
Don't check that event is set when we don't need to

baby-bell committed:
Remove HibikeMessage.__eq__

baby-bell committed:
Do not special-case single arg functions

Instead of manually checking whether our random data generator function returns a tuple to see if it generates multiple arguments, we can require that it returns an iterable instead.

baby-bell committed:
More descriptive message

baby-bell committed:
Address Yizhe concerns

kevinji commented on hibike/virtual_device.py:
```diff
@@ -89,19 +105,22 @@ def main():
             # and params, then send a subscription response
-            print("Subscription request recieved")
+            if args.verbose:
+                print("Subscription request recieved")
```
s/recieved/received

kevinji commented on hibike/virtual_device.py:
```diff
@@ -89,19 +105,22 @@ def main():
             # Send a subscription response
-            print("Ping recieved")
+            if args.verbose:
+                print("Ping recieved")
```
Ditto.

kevinji commented on hibike/virtual_device.py:
```diff
@@ -117,7 +136,8 @@ def main():
             # and return the values of the parameters written to using a device data
-            print("Device write recieved")
+            if args.verbose:
+                print("Device write recieved")
```
Ditto.

kevinji commented on hibike/virtual_device.py:
```diff
@@ -89,19 +105,22 @@ def main():
             # Send a device data with the requested param and value tuples
-            print("Device read recieved")
+            if args.verbose:
+                print("Device read recieved")
```
Ditto.

baby-bell committed:
Appease the linter gods

baby-bell committed:
Whoops, syntax error

baby-bell committed:
I before E...

baby-bell committed:
Instructions for running tests

kevinji commented:
Can you add these tests to Travis as well? You might need to modify the `requirements.txt` to include `unittest` as a dep.

baby-bell committed:
Add unit tests to Travis

baby-bell committed:
Actually install hibike deps for Travis

baby-bell committed:
Restructure to avoid checking event twice

baby-bell committed:
Factor out general purpose functions

baby-bell committed:
Remove repeated verbosity checks

baby-bell committed:
Lint fixes

baby-bell commented:
Is there anything else that needs to be resolved, @yizow ?

yizow commented on hibike/TESTS.md:
```diff
@@ -0,0 +1,6 @@
+Tests are written using Python's `unittest` module. You can read
+more about it [here](https://docs.python.org/3/library/unittest.html).
+All unit tests are located in the `unit_tests` directory. You can run
+them yourself with `python3 -m unittest unit_tests/*.py`.
```
from what directory/

yizow commented on hibike/hibike_message.py:
```diff
@@ -6,6 +6,8 @@
 import os
 import json
+import threading
+
```
extra new line. remove

yizow commented on hibike/hibike_message.py:
```diff
@@ -405,19 +407,21 @@ def parse_bytes(msg_bytes):
 
 
-def blocking_read_generator(serial_conn, stop_event=None):
+def blocking_read_generator(serial_conn, stop_event=threading.Event()):
```
Interesting quirk with python arg defaults: 
One object is created when the module is imported, and that is used for every single function call. 

So when one threading.Event() for a blocking_read_generator called with default arguments triggers, every single blocking_read_generator called with default arguments will trigger. 

I recommend keeping it None, then have an `if stop_event == None` etc.

baby-bell replied:
This shouldn't be an issue, because the event is never triggered within `blocking_read_generator`,

yizow commented on hibike/hibike_message.py:
```diff
@@ -405,19 +407,21 @@ def parse_bytes(msg_bytes):
+    # Switch to nonblocking mode so that we don't get stuck reading
+    old_timeout = serial_conn.timeout
+    serial_conn.timeout = 0
+    while not stop_event.is_set():
```
will this cause a busy wait?

baby-bell replied:
If there is no data coming down the serial pipeline, then yes. There isn't really a way to fix this without reworking Hibike's design.

yizow commented on hibike/hibike_message.py:
```diff
@@ -405,19 +407,21 @@ def parse_bytes(msg_bytes):
-
+    # Switch to nonblocking mode so that we don't get stuck reading
+    old_timeout = serial_conn.timeout
+    serial_conn.timeout = 0
```
Is there really a point to manipulating the timeout like this? Who else would change it/ when do we ever want a timeout?

baby-bell replied:
It's useful if you want to do blocking reads, then `timeout = None`. I don't want to add more preconditions.

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+"""
+Unit tests for functions in hibike_message.
```
maybe change the name of this directory to `hibike_tests` or something similar? I feel like the current name may cause namespace confusion (from the programmer's perspective)

baby-bell replied:
Done

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+import hibike_message
+import spawn_virtual_devices
+from unit_tests.utils import run_with_random_data
+DEVICE_TYPES = list(hibike_message.DEVICES)
```
add new line between imports and object initializations.

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+def random_uid():
+    """
+    Generate a random UID.
```
Pep style guide says single line should have quotes on same line
[here](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings)

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+def random_params(device_id):
+    """
+    Get a random sampling of parameters for a particular DEVICE_TYPE.
```
see above

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+def random_values(device_id, params):
+    """
+    Generate random valid values for PARAMS.
```
see above

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+            if param_pack["name"] == param:
+                param_index = idx
+                break
+        param_type = all_params[param_index]["type"]
```
I'd add an if statement here: 
`if param_index == 0`
better to be explicit, and more readable
add a corresponding `else` case that should never happen, and just throw a generic exception

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+class CobsTests(unittest.TestCase):
+    """
+    Tests for COBS encoding and decoding.
```
See docstring comment above

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+class FakeSerialPort(object):
+    """
+    A fake serial port that acts as a queue.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+    def write(self, byte_buf):
+        """
+        Write BYTE_BUF into the queue.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+    def read(self, length=1):
+        """
+        Read out LENGTH bytes from the queue.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+    def drain(self):
+        """
+        Read the contents of the queue into a bytearray.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @property
+    def in_waiting(self):
+        """
+        The number of bytes in the queue.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+class ParamsTests(unittest.TestCase):
+    """
+    Tests for encoding and decoding device parameters.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+class ParsingTests(unittest.TestCase):
+    """
+    Tests for parsing Hibike messages.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def encode_packet(packet):
+        """
+        Turn a HibikeMessage into an encoded packet.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_sub_request():
+        """
+        Generate a random subscription request packet.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+        return fake_port.drain()
+
+    @staticmethod
+    def gen_random_sub_request():
```
would it perhaps be useful to make these family of functions take arguments, with default values set to something? (maybe `None`, though if you did that, to get the same cleanless you'd have to create another function to wrap the `if None, else random()` check)

baby-bell replied:
Hm, not quite sure why this would be better. More flexibility?

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_sub_response():
+        """
+        Generate a random subscription response packet.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_device_read():
+        """
+        Generate a random device read packet.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_device_write():
+        """
+        Generate a random device write packet.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_device_data():
+        """
+        Generate a random device data packet.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_heartbeat_response():
+        """
+        Generate a heartbeat response with a random ID.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+    @staticmethod
+    def gen_random_error():
+        """
+        Generate an error packet with a random error code.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+    def test_parse_bad_checksum(self):
+        """
+        Packets with bad checksums should not be parsed.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+        def assert_parse_is_none(invalid_packet):
+            """
+            Assert that parsing an invalid packet results in None.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+    def test_parse_idempotence(self):
+        """
+        Check that parsing an encoded packet is idempotent.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+class BlockingReadGeneratorTests(unittest.TestCase):
+    """
+    Tests for blocking_read_generator.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_message_tests.py:
```diff
@@ -0,0 +1,373 @@
+
+    def test_read_device_data(self):
+        """
+        Check that DeviceData packets are received and decoded.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_process_tests.py:
```diff
@@ -0,0 +1,65 @@
+
+    def test_detect_devices(self):
+        """
+        Test detection of valid devices.
```
See docstring comment

yizow commented on hibike/unit_tests/hibike_process_tests.py:
```diff
@@ -0,0 +1,65 @@
+
+    def test_detect_no_devices(self):
+        """
+        Make sure that we don't detect empty serial ports as sensors.
```
See docstring comment

yizow commented on hibike/virtual_device.py:
```diff
@@ -26,11 +26,22 @@ def main():
+                        action="store_true")
     args = parser.parse_args()
 
+    def verbose_log(fmt_string, *fmt_args):
```
I like this.
we should think about pulling this out into some general utilities shared between hibike + runtime. although that may be too much effort.

baby-bell committed:
Rename unit_tests/ to hibike_tests/

baby-bell committed:
Update TESTS to reflect new directory

baby-bell committed:
Fix docstring comments

baby-bell committed:
Remove extra newline

baby-bell committed:
Reflect changed test dir in travis Makefile

baby-bell commented:
Okay, I think that's pretty much everything. Is there anything else that needs to be resolved, @yizow? I appreciate all the constructive criticism so far.

baby-bell committed:
Merge branch 'master' into hibike/unit-tests

baby-bell committed:
Merge branch 'master' of github.com:pioneers/PieCentral into hibike/unit-tests

Check to make sure latest branch passes tests
